t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -562,7 +562,12 @@ test_expect_success 'reset with sparse directory pathspec outside definition' '
 	test_all_match git status --porcelain=v2
 '
 
-test_expect_success 'reset with file pathspec outside sparse definition' '
+# TODO: reset with a file pathspec outside of the sparse checkout definition
+# does not currently work in cone-mode checkouts. This is due to the tree's
+# directory entry (that contains the file) not being expanded before the
+# pathspec check is performed (in the case of this test, "folder1/" is compared
+# to the pathspec "folder1/a", and does not match).
```

Comment:
Thanks for pointing out how this is failing. We should debug into this tomorrow to see what logic is happening. Something like `index_name_pos(istate, "folder1/a")` would expand a sparse index, so something different must be happening. Perhaps it's related to how `reset` uses the cache-tree structure.

Reply:
The fact that we are getting into `oneway_diff()` with a `folder1/` entry an a `folder1/a` entry makes me think that we need more updates to `unpack_trees()` to catch this case during that update. When you debug into that line, is there anything in the stack trace that looks like a good place to identify that we need to expand the index here?

As for preventing the user request, there is some precedent in `builtin/add.c` using `matches_skip_worktree()` and `path_in_sparse_checkout()` that might be illuminating. Take a look and see if you are inspired for an approach in `git reset`. There is absolutely momentum building to prevent these kinds of interactions outside of the sparse-checkout cone.

Reply:
I've just added two fixups (for readability in this PR - I'll squash them before merging) that should fix this. The first adds more tests for pathspec resets, and the second adds checks to expand the index if the pathspec isn't obviously for things in the index (and corresponding tests for when the index should _not_ be expanded).